OSSM 3.0 TP1 OSSM-8153 [DOC] OSSM 3.x Concepts (#83560)
openshift/openshift-docs: modules/ossm-about-concepts-resources.adoc

// Module included in the following assemblies:
// about/ossm-about-concepts-assembly.adoc

:_mod-docs-content-type: CONCEPT
[id="ossm-about-concepts-resources_{context}"]
= {SMProductName} resources

{SMProductName} Operator manages the lifecycle of your Istio control planes. Instead of creating a new configuration schema, {SMProduct} Operator APIs are built around Istio's Helm chart APIs.

//All installation and configuration options that are exposed by Istio's Helm charts are available through the {SMProduct} Operator Custom Resource Definition (CRD) `values` fields.

[NOTE]
====
* Though {SMProductName} APIs are built around Istio's Helm chart APIs, Helm charts are not supported.
* All installation and configuration options that are exposed by Istio's Helm charts are available through the {SMProductName} Custom Resource Definition (CRD) `values` fields.
====

[id="istio-resource_{context}"]
== Istio resource

> The `Istio` resource is used to manage your Istio control planes. It is a cluster-wide resource, as the Istio control plane operates in and requires access to the entire cluster.

Reviewer: Per [ISG](https://www.ibm.com/docs/en/ibm-style?topic=word-usage#a): Do not use "as" to mean because or while.

```suggestion
The `Istio` resource is used to manage your Istio control planes. It is a cluster-wide resource, because the Istio control plane operates in and requires access to the entire cluster.
```

> To select a `namespace` to run the control plane pods in, you can use the `spec.namespace` field.

Reviewer: Don't think we need mark-up here, as we are talking about namespaces generally, not the namespace object.

```suggestion
To select a namespace to run the control plane pods in, you can use the `spec.namespace` field.
```

> * CRDs are supported.

Reviewer 2: What CRDs?

Author: Custom Resource Defintion

Reviewer 2: What CRDs are supported? The Istio CRDs? I'm not sure specifically _what_ CRDs are supported here.

Author: @longmuir , are there other CRDs that are supported, or just OpenShift Service Mesh CRDs?

Reviewer 3: I'm not sure what the last bullet here "*CRDs are supported" refers to...it looks as if it's missing something before it. There are a lot of CRDs we support under "OpenShift Service Mesh CRDs"... the sail operator ones (Istio, IstioRevision), Istio's CRDs (VirtualService, etc),.. even Kiali's CRDs (Kiali, OSSMConsole). 

It's on us to call out what we don't support. Probably the only one I can think of that we don't support directly is the upstream `IstioOperator` resource, as we have the `Istio` resource (though I believe this is captured).

Author: I can take that last bullet point out.

[NOTE]
====

> The `spec.namespace` field is immutable: in order to move a control plane to another namespace, you have to remove the `Istio` resource and recreate it with a different `spec.namespace`.

Reviewer: Per [ISG](https://www.ibm.com/docs/en/ibm-style?topic=word-usage#h): Use "have to" only when must is too restrictive 

```suggestion
The `spec.namespace` field is immutable: in order to move a control plane to another namespace, you must remove the `Istio` resource and recreate it with a different `spec.namespace`.
```

====

> You can access all `Istio` custom resource defintion (CRD) options through `spec.values` fields:

Reviewer 4: ```suggestion
You can access all `Istio` custom resource definition (CRD) options through `spec.values` fields:
```

> To select a `namespace` to run the control plane pods in, you can use the `spec.namespace` field. `.

Reviewer 2: ```suggestion
To select a `namespace` to run the control plane pods in, you can use the `spec.namespace` field.
```

The `spec.namespace` field is immutable: in order to move a control plane to another namespace, you have to remove the `Istio` resource and recreate it with a different `spec.namespace`.
====

> You can access all `Istio` custom resource defintion (CRD) options through `spec.values` fields:

Reviewer 2: I think it'd be worth mentioning here that you can use `oc explain istios.spec.values` to see all the customization options.

apiVersion: sailoperator.io/v1alpha1
kind: Istio
metadata:
  name: default
spec:
  version: v1.22.3
  namespace: istio-system
  updateStrategy:
    type: InPlace
  values:
    pilot:
      resources:
        requests:
          cpu: 100m
          memory: 1024Mi
----

You can run the following command to see all the customization options:

[source, terminal]
----
$ oc explain istios.spec.values
----
//Helm charts can only be used with istio-csr and Helm charts are a temporary work around for istio-csr. This might get confusing for users, so a NOTE was added.

> To support canary updates of the control plane, {SMProduct} includes support for multiple Istio versions. You can select a version by setting  `spec.version` to the version you would like to install, prefixed with a v. You can update to a new version just by changing this field.

Reviewer: Mark-up here, as a value?
```suggestion
To support canary updates of the control plane, {SMProduct} includes support for multiple Istio versions. You can select a version by setting  `spec.version` to the version you would like to install, prefixed with a `v`. You can update to a new version just by changing this field.
```

> Istio uses a `ConfigMap` for its global configuration, called the `MeshConfig`. All of its settings are available through `spec.meshConfig`.

Reviewer 2: This heading doesn't make sense to me. Only 1 sentence talks about `MeshConfig` but the rest is about upgrades and update strategy?

{SMProduct} supports two different update strategies for your control planes:

> `InPlace`:: {SMProduct} Operator immediately replaces your existing control plane resources with the ones for the new version.

Reviewer: ```suggestion
`InPlace`:: The {SMProduct} Operator immediately replaces your existing control plane resources with the ones for the new version.
```

`RevisionBased`:: Uses Istio's canary update mechanism by creating a second control plane to which you can migrate your workloads to complete the update.

After creating an Istio resource, {SMProduct} generates a revision name for the resource based on the `updateStrategy`, and creates a corresponding `IstioRevision`.

[id="istiorevision-resource_{context}"]
== IstioRevision resource

> The `IstioRevision` is a cluster-wide resource and the lowest-level API {SMProduct} provides. It is usually not created by the user, but by the operator itself. Its schema closely resembles that of the `Istio` resource - but instead of representing the state of a control plane you want to be present in your cluster, it represents a revision of that control plane.

Reviewer: ```suggestion
The `IstioRevision` is a cluster-wide resource and the lowest-level API {SMProduct} provides. It is usually not created by the user, but by the Operator itself. Its schema closely resembles that of the `Istio` resource - but instead of representing the state of a control plane you want to be present in your cluster, it represents a revision of that control plane.
```

Do we need any mention of when/why the resource wouldn't be created by the Operator?

Author: Not for TP1.

A revision of the control plane you want to be present in your cluster is an instance of Istio with a specific version and revision name, and its revision name can be used to add workloads or entire namespaces to the mesh. For example: by using the `istio.io/rev=<REVISION_NAME>` label.

> You can think of the relationship between the `Istio` and `IstioRevision` resources as similar to the relationship between Kubernetes' `ReplicaSet` and Pod: a `ReplicaSet` can be created by users and results in the automatic creation of Pods, which will trigger the instantiation of your containers.

Reviewer: ```suggestion
You can think of the relationship between the `Istio` and `IstioRevision` resources as similar to the relationship between Kubernetes' replica set and pod: a replica set can be created by users and results in the automatic creation of pods, which will trigger the instantiation of your containers.
```

> Similarly, users create an `Istio` resource which instructs the {SMProduct} Operator to create a matching `IstioRevision`, which then in turn triggers the creation of the Istio control plane. To do that, the {SMProduct} Operator will copy all of your relevant configuration from the `Istio` resource to the `IstioRevision` resource.

Reviewer: ```suggestion
Similarly, users create an `Istio` resource which instructs the {SMProduct} Operator to create a matching `IstioRevision` object, which then in turn triggers the creation of the Istio control plane. To do that, the {SMProduct} Operator will copy all of your relevant configuration from the `Istio` resource to the `IstioRevision` resource.
```

> Similarly, users create an `Istio` resource which instructs the {SMProduct} Operator to create a matching `IstioRevision`, which then in turn triggers the creation of the Istio control plane. To do that, the {SMProduct} Operator will copy all of your relevant configuration from the `Istio` resource to the `IstioRevision` resource.

Reviewer 2: This section does a good job of explaining how the two resources relate to each other!

[id="istiocni-resource_{context}"]
== IstioCNI resource

The lifecycle of Istio's Container Network Interface (CNI) plugin is managed separately when using {SMProduct} Operator. To install Istio's CNI plugin, you create an `IstioCNI` resource.

> The `IstioCNI` resource is a cluster-wide resource as it installs a `DaemonSet` that operates on all nodes of your cluster. You can select a version by setting the `spec.version` field, as you can see in the example below. To update the CNI plugin, change the version field to the version you want to install. Just like the `Istio` resource, it also has a `values` field that exposes all of the options provided in the `istio-cni` chart:

Reviewer: Per [ISG](https://www.ibm.com/docs/en/ibm-style?topic=word-usage#b): Do not use to indicate a relative location in a document
```suggestion
The `IstioCNI` resource is a cluster-wide resource as it installs a daemon set that operates on all nodes of your cluster. You can select a version by setting the `spec.version` field, as you can see in the example that follows. To update the CNI plugin, change the version field to the version you want to install. Just like the `Istio` resource, it also has a `values` field that exposes all of the options provided in the `istio-cni` chart:
```

* CRDs are supported.
====

[id="istio-resource_{context}"]
== `Istio` resource

The `Istio` resource is used to manage your Istio control planes. It is a cluster-wide resource, as the Istio control plane operates in and requires access to the entire cluster.

To select a `namespace` to run the control plane pods in, you can use the `spec.namespace` field.

[NOTE]
====
The `spec.namespace` field is immutable: in order to move a control plane to another namespace, you have to remove the `Istio` resource and recreate it with a different `spec.namespace`.
====

You can access all `Istio` custom resource defintion (CRD) options through `spec.values` fields:

.Example `Istio` resource CRD
[source,yaml]
----
apiVersion: sailoperator.io/v1alpha1
kind: Istio
metadata:
  name: default
spec:
  version: v1.22.3
  namespace: istio-system
  updateStrategy:
    type: InPlace
  values:
    pilot:
      resources:
        requests:
          cpu: 100m
          memory: 1024Mi
----

You can run the following command to see all the customization options:

> +

Reviewer 2: Extra `+` here.
```suggestion
```
